File: interviewPrep/design_principles/SOLID/LiskovSubstitutionPrinciple/typescript/passLiskovSubstitutionPrinciple.ts
interface FlyingBird {
  fly(): void;
}

interface SwimmingBird {
  swim(): void;
}

class Duck2 implements FlyingBird {
  quack() {
    console.log("I can quack");
  }

  fly() {
    console.log("I can fly");
  }
}

class Penguin2 implements SwimmingBird {
  swim() {
    console.log("I can swim");
  }
}

function makeFlyingBirdFly(bird: FlyingBird) {
  bird.fly();
}

function makeSwimmingBirdSwim(bird: SwimmingBird) {
  bird.swim();
}

const duck2 = new Duck2();
const penguin2 = new Penguin2();

makeFlyingBirdFly(duck2);
makeSwimmingBirdSwim(penguin2);

File: interviewPrep/design_principles/SOLID/LiskovSubstitutionPrinciple/python/passLiskovSubstitutionPrinciple.py
class FlyingBird:
    def fly(self):
        print("I can fly")


class SwimmingBird:
    def swim(self):
        print("I can swim")


class Duck(FlyingBird):
    def quack(self):
        print("I can quack")


class Penguin(SwimmingBird):
    ...


def make_flying_bird_fly(bird):
    bird.fly()


def make_swimming_bird_swim(bird):
    bird.swim()


duck = Duck()
penguin = Penguin()

make_flying_bird_fly(duck)
make_swimming_bird_swim(penguin)
